system_setup/server/controllers/locale.py

```diff
@@ -30,3 +32,65 @@ def start(self):
                           .format(self.model.selected_language))
 
         super().start()
+
+    async def _run_locale_support_cmds(self, lang: str):
+        """ Final commands to run when a valid language is POST'ed. """
+        cmds = [["locale-gen"],
+                ["update-locale", "LANG={}".format(lang)]]
+        env = os.environ.copy()
+
+        cp = await arun_command(["check-language-support", "-l", lang[0:5]],
+                                env=env)
+        if cp.returncode == 0:
+            packages = cp.stdout.strip('\n').split(' ')
+            cmds.append(["apt", "install", "-y"] + packages)
+
+        if self.app.opts.dry_run:
+            for cmd in cmds:
+                log.debug('Would run: ' + ' '.join(cmd))
+        else:
+            for cmd in cmds:
+                cp = await arun_command(cmd, env=env)
+                if cp.returncode:
+                    log.debug('Command \"%s\" failed with return code %d',
+                              cp.args, cp.returncode)
+
+    async def POST(self, data: str):
+        if data == self.autoinstall_default or data == os.environ.get("LANG"):
+            await super().POST(data)
+            return
+
+        fileContents: str
+        fname = "locale.gen"
+        env = os.environ.copy()
+        if self.app.opts.dry_run:
+            # For debug purposes.
+            fname = ".subiquity/" + fname
+            await arun_command(['cp', '/etc/locale.gen', '.subiquity/'],
+                               env=env)
+        else:
+            fname = "/etc/" + fname
+
+        pendingWrite = False
+        with open(fname, "r") as localeGen:
+            # locale.gen is not so big.
+            fileContents = localeGen.read()
+            lineFound = fileContents.find(data)
+            if lineFound == -1:
+                # An unsupported locale coming from our UI is a bug.
+                raise AssertionError("Selected language {} not supported."
+                                     " Rolling back.".format(data))
+
+            commented = "# {}".format(data)
+            lineFound = fileContents.find(commented)
+            if lineFound != -1:
+                fileContents = fileContents.replace(commented, data)
+                pendingWrite = True
+
+        if pendingWrite:
+            with open(fname, "wt") as f:
+                f.write(fileContents)
```

Comment:
what if we can’t write it? (and yes, I know, there are a lot of problems like this in the code), but the exception wouldn’t be posted back to the caller if I’m correct

Reply:
I guess you're correct, because I've seen TUI gettng stuck on the failure report screen when server is down due uncaught exceptions. Put into a try-except block and log.error.

Reply:
+ return False.